src/repository.rs

```diff
@@ -537,7 +538,318 @@ where
+    /// stored on the repository. If `None`, this defaults to `/`. For example, if there is a TUF
+    /// repository at `https://tuf.example.com/`, but all targets are stored at `/targets/`, then
+    /// passing the arg `Some("targets".into())` would cause `hello-world` to be fetched from
+    /// `https://tuf.example.com/targets/hello-world`.
+    pub fn targets_prefix(mut self, targets_prefix: Vec<String>) -> Self {
+        self.targets_prefix = Some(targets_prefix);
+        self
+    }
+
+    /// Set the minimum bytes per second for a read to be considered good.
+    pub fn min_bytes_per_second(mut self, min: u32) -> Self {
+        self.min_bytes_per_second = min;
+        self
+    }
+
+    /// Build a `HttpRepository2`.
+    pub fn build(self) -> HttpRepository2<C, D> {
+        let user_agent = match self.user_agent {
+            Some(user_agent) => user_agent,
+            None => "rust-tuf".into(),
+        };
+
+        HttpRepository2 {
+            uri: self.uri,
+            client: self.client,
+            interchange: self.interchange,
+            user_agent: user_agent,
+            metadata_prefix: self.metadata_prefix,
+            targets_prefix: self.targets_prefix,
+            min_bytes_per_second: self.min_bytes_per_second,
+        }
+    }
+}
+
+/// A repository accessible over HTTP.
+pub struct HttpRepository2<C, D>
+where
+    C: Connect + Sync + 'static,
+    D: DataInterchange,
+{
+    uri: Uri,
+    client: Client<C>,
+    user_agent: String,
+    metadata_prefix: Option<Vec<String>>,
+    targets_prefix: Option<Vec<String>>,
+    min_bytes_per_second: u32,
+    interchange: PhantomData<D>,
+}
+
+// Configuration for urlencoding URI path elements.
+// From https://url.spec.whatwg.org/#path-percent-encode-set
+const URLENCODE_FRAGMENT: &percent_encoding::AsciiSet = &percent_encoding::CONTROLS
+    .add(b' ')
+    .add(b'"')
+    .add(b'<')
+    .add(b'>')
+    .add(b'`');
+const URLENCODE_PATH: &percent_encoding::AsciiSet = &URLENCODE_FRAGMENT
+    .add(b'/')
+    .add(b':')
+    .add(b';')
+    .add(b'=')
+    .add(b'@')
+    .add(b'[')
+    .add(b'\\')
+    .add(b']')
+    .add(b'^')
+    .add(b'|');
+
+fn extend_uri(uri: Uri, prefix: &Option<Vec<String>>, components: &[String]) -> Result<Uri> {
+    let mut uri_parts = uri.into_parts();
+
+    let (path, query) = match &uri_parts.path_and_query {
+        Some(path_and_query) => (path_and_query.path(), path_and_query.query()),
+        None => ("", None),
+    };
+
+    let mut modified_path = path.to_owned();
+    if modified_path.ends_with("/") {
+        modified_path.pop();
+    }
+
+    let mut path_split = modified_path
+        .split("/")
+        .map(String::from)
+        .collect::<Vec<_>>();
+    let mut new_path_elements: Vec<String> = vec![];
+
+    if let Some(ref prefix) = prefix {
+        new_path_elements.extend(prefix.iter().cloned());
+    }
+    new_path_elements.append(&mut components.to_vec());
```

Comment:
`to_vec` copies the list, but you should be able to just use the slice directly by using `extend_from_slice` or `extend` instead of `append`.

Reply:
Done.